go/tools/builders/cgo.go

```diff
@@ -0,0 +1,175 @@
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// cgo invokes "go tool cgo" after filtering souces, and then process the output
+// into a normalised form. It is invoked by the Go rules as an action.
+package main
+
+import (
+	"flag"
+	"fmt"
+	"go/ast"
+	"go/build"
+	"go/parser"
+	"go/token"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+)
+
+func testCgo(src string, data []byte) (bool, string, error) {
+	fset := token.NewFileSet()
+	parsed, err := parser.ParseFile(fset, src, data, parser.ImportsOnly)
+	if err != nil {
+		return false, "", err
+	}
+	for _, decl := range parsed.Decls {
+		d, ok := decl.(*ast.GenDecl)
+		if !ok {
+			continue
+		}
+		for _, dspec := range d.Specs {
+			spec, ok := dspec.(*ast.ImportSpec)
+			if !ok {
+				continue
+			}
+			imp, err := strconv.Unquote(spec.Path.Value)
+			if err != nil {
+				log.Panicf("%s: invalid string `%s`", src, spec.Path.Value)
+			}
+			if imp == "C" {
+				return true, parsed.Name.String(), nil
+			}
+		}
+	}
+	return false, parsed.Name.String(), nil
+}
+
+func run(args []string) error {
+	sources := multiFlag{}
+	cc := ""
+	objdir := ""
+	flags := flag.NewFlagSet("cgo", flag.ContinueOnError)
+	flags.Var(&sources, "src", "A source file to be filtered and compiled")
+	flags.StringVar(&cc, "cc", "", "Sets the c compiler to use")
+	flags.StringVar(&objdir, "objdir", "", "The output directory")
+	// process the args
+	if len(args) < 2 {
+		flags.Usage()
+		return fmt.Errorf("The go tool must be specified")
+	}
+	gotool := args[0]
+	if err := flags.Parse(args[1:]); err != nil {
+		return err
+	}
+
+	// apply build constraints to the source list
+	// also pick out the cgo sources
+	bctx := build.Default
+	bctx.CgoEnabled = true
+	cgo_srcs := []string{}
+	for _, s := range sources {
+		bits := strings.SplitN(s, "=", 2)
+		if len(bits) != 2 {
+			return fmt.Errorf("Invalid source arg, expected output=input got %s", s)
+		}
+		out := bits[0]
+		in := bits[1]
+		// Check if the file is filtered first
+		data, err := ioutil.ReadFile(in)
+		if err != nil {
+			return err
+		}
+		match, err := matchFile(bctx, in)
+		if err != nil {
+			return err
+		}
+		// if this is not a go file, it cannot be cgo, so just check the filter
+		if !strings.HasSuffix(in, ".go") {
+			// Not a go file, just filter, assume C or C-like
+			if match {
+				// not filtered, copy over
+				if err := ioutil.WriteFile(out, data, 0644); err != nil {
+					return err
+				}
+			} else {
+				// filtered, make empty file
+				if err := ioutil.WriteFile(out, []byte("//Empty file"), 0644); err != nil {
```

Comment:
I would actually leave these files empty. // may not be a valid comment prefix for assembly dialects compiled by the external compiler.